### PR TITLE
Refactor NLLLoss to support split class dim

### PR DESCRIPTION
**Pull request by leaves-zwx**

### CODE REVIEW

File: oneflow/core/functional/impl/nn_functor.cpp [hunk 1]
```diff
@@ -953,39 +950,70 @@ class NllLossFunctor {
         << "Reduction should be none, sum or mean.";
 
     const auto& input_shape = input->shape();
+    const int64_t K = input_shape->NumAxes();
+    CHECK_GE_OR_RETURN(K, 2) << "Expected 2 or more dimensions";
```
Comment (MARD1NO) [inline oneflow/core/functional/impl/nn_functor.cpp:954]:
```cpp
CHECK_GE_OR_RETURN(K, 2) << Error::RuntimeError() << "Expected 2 or more dimensions";
```
下同

File: oneflow/user/ops/nll_op.cpp [hunk 2]
```diff
@@ -14,125 +14,188 @@ See the License for the specific language governing permissions and
...
+  return Maybe<void>::Ok();
 }
 
-/* static */ Maybe<void> NllOp::ModifyInputArg(const GetInputArgModifier& GetInputArgModifierFn,
+/* static */ Maybe<void> NLLOp::ModifyInputArg(const GetInputArgModifier& GetInputArgModifierFn,
                                                const user_op::UserOpConfWrapper& conf) {
   user_op::InputArgModifier* target_modifier = GetInputArgModifierFn("target", 0);
   CHECK_OR_RETURN(target_modifier != nullptr);
   target_modifier->set_requires_grad(false);
+  auto* weight_modifier = GetInputArgModifierFn("weight", 0);
```
Comment (MARD1NO) [inline oneflow/user/ops/nll_op.cpp:114]: 这里有没有必要判断下 has_input

**Review by MARD1NO — APPROVED** [review]

File: oneflow/user/ops/nll_op.cpp (hunk 2 above)
Comment (leaves-zwx) [reply oneflow/user/ops/nll_op.cpp:114]: 我看了下实现，里面是 find arg name，如果 find 不到会直接返回 nullptr，所以下文直接判断了 weight_modifier 是不是为空。

**Review by mosout — APPROVED** [review]

File: oneflow/user/ops/nll_op.cpp [hunk 3]
```diff
@@ -14,125 +14,188 @@ See the License for the specific language governing permissions and
...
+  // split batch dim
+  auto builder1 = ctx->NewBuilder()
+                      .Split(user_op::OpArg("input", 0), 0)
+                      .Split(user_op::OpArg("target", 0), 0)
+                      .Split(user_op::OpArg("output", 0), 0);
+  if (ctx->user_op_conf().has_input("weight", 0)) {
+    builder1.Broadcast(user_op::OpArg("weight", 0));
+  }
+  if (ctx->user_op_conf().has_output("total_weight", 0)) {
+    builder1.Broadcast(user_op::OpArg("total_weight", 0));
```
Comment (guo-ran) [inline oneflow/user/ops/nll_op.cpp:88]: 看这个total_weight是sum，这个情况下是不是也是P

Comment (leaves-zwx) [reply oneflow/user/ops/nll_op.cpp:88]: 哦，对。如果我这里输出一个 shape 为 [N, ] 的 out_weight 是不是更好 ？
sum 的工作可以在外面做，而且可以避免在 nll 的 cuda kernel 使用 atomic_add ?

Comment (guo-ran) [reply oneflow/user/ops/nll_op.cpp:88]: 是的，而且half的atomic_add巨慢

Comment (guo-ran) [reply oneflow/user/ops/nll_op.cpp:88]: 嗯，reduce sum里应该没用atomic_add

**Comment by leaves-zwx** [comment]
现在在单元测试 (test_nll_loss.py) 中遇到1个问题，test_2d_parallel_weighted 会几率失败。根据调试，很像 graph 反向时 weight 的内存被非法修改了。

**Comment by leaves-zwx** [comment]
单测只在 device 为 cpu 时有几率失败，反向的输入 weight 和 target 都没问题，每次失败时，都是 out_grad 的数值不正确，需要往前看是不是其他 op 的反向有问题。

**Comment by leaves-zwx** [comment]
在与 @BBuf 沟通后发现，autotest 的 check_graph=True 时是不会跑 backward 的，但报错的是 grad 的数值不对，那也就是说是 eager 下的 grad 的数值不对。但 check_graph=False 后，运行很多次 (50) 都不会报错，而 check_graph=True 下 < 5 次必然出错。

这里有理由怀疑 autotest 下面 graph 与 eager 的数据有污染，或者是某种错误用法造成的（比如缺少 detach），或者某种底层机制造成的（非预期的内存共享）。

File: oneflow/user/kernels/nll_kernel_util.cu [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  const T zero = GetZeroVal<T>();
+  const T one = GetOneVal<T>();
+  CUDA_1D_KERNEL_LOOP(i, num_samples) {
+    K label = target[i];
+    if (label != ignore_index) {
+      label -= class_start;
+      if (label >= 0 && label < num_classes) {
+        const T w = weight ? weight[label] : one;
+        out[i] = -(input[i * num_classes + label] * w);
+        if (out_weight) { out_weight[i] = w; }
```
Comment (guo-ran) [inline oneflow/user/kernels/nll_kernel_util.cu:36]: 这里感觉有问题：
1、有out_weight时一定有weight吗？如果没有那此时w是one，out_weight赋值为1是否合适。
2、如果label不在(label >= 0 && label < num_classes)范围内，out_weight应该设0吧。

Comment (leaves-zwx) [reply oneflow/user/kernels/nll_kernel_util.cu:36]: 1. kernel 里面应该不关心 weight 和 out_weight 的绑定（同时有或同时没有），根据 out_weight 的使用场景，如果 weight 为 null，那么每个元素的 out_weight 都是1也没问题。
2. 这里好像有问题，好像会造成 undefined behavior 的错误。

Comment (guo-ran) [reply oneflow/user/kernels/nll_kernel_util.cu:36]: 如果 weight 为 null，得到的out_weight最后会是parallel_num，因为sbp是P，会累加起来。

Comment (leaves-zwx) [reply oneflow/user/kernels/nll_kernel_util.cu:36]: image

Comment (leaves-zwx) [reply oneflow/user/kernels/nll_kernel_util.cu:36]: weight shape 为 (C,)
out_weight shape 为 (N,)

如果 C 维被切分，那 input sbp 是 S(1)，weight 如果存在，sbp 就是 S(0)，out_weight 如果存在，sbp 是 B（因为 N 不切）。

为什么说 out_weight 为 P 呢？

而且在 op 层面 weight 和 out_weight 是同时存在的。

Comment (guo-ran) [reply oneflow/user/kernels/nll_kernel_util.cu:36]: 我这里指的是C维被切分的情况，即S1，我看你的输出out_weight是(N,)，那此时out_weight就应该是P吧

Comment (leaves-zwx) [reply oneflow/user/kernels/nll_kernel_util.cu:36]: 哦哦，我说错了，out_weight 就是 P

Comment (leaves-zwx) [reply oneflow/user/kernels/nll_kernel_util.cu:36]: 根据你的提示修改后，再测试就没问题了

**Comment by leaves-zwx** [comment]
测试已经通过，是 kernel 本身的问题

**Comment by leaves-zwx** [comment]
在随机测试中遇到1个问题，当 weight 全部为 0，或者 target 中随机的 label 全部都等于 ignore_index 时，reduction 为 mean 时，最后需要除 total weight，但此时 total weight 为 0，pytorch 给出的结果为 0，而 oneflow 直接返回 nan。

**Comment by leaves-zwx** [comment]
有1个 corner case 没完全解决，https://github.com/Oneflow-Inc/OneTeam/issues/1497
